### PR TITLE
ChIP-seq pipeline: removing reads in blacklist regions

### PR DESCRIPTION
**Pull request by orzechoj**
Added a module to remove read in blacklist regions (e.g. from ENCODE), which uses bedtools intersect.
(Also raised the memory for cf_merge)

### CODE REVIEW

**Comment by ewels** [comment]
Super cool! Any thoughts on using the genome reference config files to store the location for the blacklist file? Presumably it will always be the same for each species? So it could have its own reference type? _eg:_

```
@reference blacklist	GRCh37	/path/to/genomes/Human/GRCh37/blacklist/	Human	GRCh37
@reference blacklist	GRCm38	/path/to/genomes/Mouse/GRCm38/blacklist/	Mouse	GRCm38
```

Might be better than having to specify it as a `param`?

I've had this functionality in the back of my mind for a while now anyway, could be good in other pipelines too.. Thanks!

**Comment by orzechoj** [comment]
Hi,

Haven’t thought much about this..

I guess black list files don’t change much for an individual genome. But I
could also see cases where you might use this module to remove reads using
other files, e.g from a few manually curated regions, from everything
overlapping lncRNAvor something else.

But as long as it’s possible to use params to set other “black list” files,
it might be a good a idea to have a default option in the config.

cheers,
Jakub

On Tuesday, April 19, 2016, Phil Ewels <notifications@github.com> wrote:
> Super cool! Any thoughts on using the genome reference config files to
store the location for the blacklist file? Presumably it will always be the
same for each species? So it could have its own reference type? eg:
>
> @reference blacklist GRCh37 /path/to/genomes/Human/GRCh37/blacklist/
Human GRCh37
> @reference blacklist GRCm38 /path/to/genomes/Mouse/GRCm38/blacklist/
Mouse GRCm38
>
> Might be better than having to specify it as a param?
>
> I've had this functionality in the back of my mind for a while now
anyway, could be good in other pipelines too.. Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub<
https://ci3.googleusercontent.com/proxy/Ryb2LFxu99ew131fAXk7zQzmZuUyqQN12R8BdJRYV6GgRg7RqltpzxZreHwxS_WTQi7YSMtjaBhskcnjY0_D5Gx16J2d3EahqPlSYAVY-3ZLn3lLhBDDdjTzMxeZkdicJpLnDCOFDKISQGOGACo3_mzqag6P9w=s0-d-e1-ft#https://github.com/notifications/beacon/AH70ibirVjfJUr3QkU2JEgeZUZ2zlBpyks5p5PNBgaJpZM4IKbeJ.gif
>

**Comment by ewels** [comment]
Yup - that could definitely work, start by looking for the a param file and if that's not found look for a genome file instead (should just be a couple of extra lines?). I just prefer to keep species specific stuff out of pipelines where possible, so that they can be used for any organism.

Phil

File: modules/bedtools_intersectNeg.cfmod.pl [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  $output_fn .= "_noblacklist.bam";
+
+
+  # Check if a file with blacklist regions was provided.
+  my $blacklistFile;
+  if(defined($cf{'params'}{'blacklistFile'})){
+      $blacklistFile = $cf{'params'}{'blacklistFile'};
+      warn "###CF bedtools intersectNeg: Using blacklist file $blacklistFile.\n";
+  } else {
+      warn "###CF bedtools intersectNeg: No blacklist file given. Skipping this step.\n";
```
Comment (ewels) [inline modules/bedtools_intersectNeg.cfmod.pl]: Nice idea to not kill the entire pipeline!

If you put the word `error` somewhere in this string, the summary script [will recognise it](https://github.com/ewels/clusterflow/blob/master/modules/cf_runs_all_finished.cfmod.pl#L139-L144) and the summary e-mail will give a prominent warning / error message about it. Nothing fancy needed, eg:

```perl
warn "###CF bedtools intersectNeg: Error - no blacklist file given. Skipping this step.\n";
```

**Comment by ewels** [comment]
Thanks again @orzechoj! I'll add the minor suggestions.

**Pull request merged**